src/mots/module.py

```diff
@@ -0,0 +1,197 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Module operations and utility functions."""
+
+from __future__ import annotations
+import logging
+from pathlib import Path
+import pprint
+
+logger = logging.getLogger(__name__)
+
+print = pprint.PrettyPrinter().pprint
+
+
+class Module:
+    """A top-level module or a submodule.
+
+    :param machine_name: a unique, machine-readable name for the module
+    :param repo_path: the path of the top-level repository
+    :param name: the name of the module
+    :param description: the description of the module
+    :param includes: a list of paths (glob format) to include
+    :param excludes: a list of paths (glob format) to exclude
+    :param owners: a list of owners that will own all paths in this module
+    :param peers: a list of peers for this module
+    :param meta: a dictionary of meta data related to this module
+    :param parent: the parent module of this module, if this is a submodule
+    :param submodules: a list of submodules of this module
+    :param exclude_submodule_paths: when True, paths in submodules are excluded
+    :param exclude_module_paths: when True, common paths in other modules are excluded
+        from the directory index
+    """
+
+    def __repr__(self):
+        """Return a human readable name for the instance."""
+        return f"<Module: {self.machine_name}>"
+
+    def __init__(
+        self,
+        machine_name: str,
+        repo_path: str,
+        name: str = None,
+        description: str = None,
```

Comment:
Small nit:
Having an `Optional[str]` has essentially three states:
* `None`
* `""`
* `"<contents>"`

For most intents and purposes, `None` == `""`, and having to track _both_ cases for validation/printing/etc can be a pain.
It might be worth considering leaning towards using `""` to represent "unset string properties" instead of `None` to simplify related logic.

Reply:
Fixed those two, and also the type hints were incorrect for `includes` and `excludes` below, which are also fixed now.